cub/test/catch2_test_device_run_length_encode_non_trivial_runs.cu

```diff
@@ -0,0 +1,251 @@
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *     * Neither the name of the NVIDIA CORPORATION nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL NVIDIA CORPORATION BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ ******************************************************************************/
+
+#include <cub/device/device_run_length_encode.cuh>
+
+#include <thrust/device_vector.h>
+#include <thrust/host_vector.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/discard_iterator.h>
+#include <thrust/logical.h>
+#include <thrust/sequence.h>
+
+#include <algorithm>
+#include <limits>
+#include <numeric>
+
+// Has to go after all cub headers. Otherwise, this test won't catch unused
+// variables in cub kernels.
+#include "catch2/catch.hpp"
+#include "catch2_test_cdp_helper.h"
+#include "catch2_test_helper.h"
+
+DECLARE_CDP_WRAPPER(cub::DeviceRunLengthEncode::NonTrivialRuns, run_length_encode);
+
+// %PARAM% TEST_CDP cdp 0:1
+
+using all_types = c2h::type_list<std::uint8_t,
+                                 std::uint64_t,
+                                 std::int8_t,
+                                 std::int64_t,
+                                 ulonglong2,
+                                 c2h::custom_type_t<c2h::equal_comparable_t>>;
+
+using types = c2h::type_list<std::uint32_t,
+                             std::int8_t>;
+
+CUB_TEST("DeviceRunLengthEncode::NonTrivialRuns can handle empty input", "[device][run_length_encode]")
+{
+  const int num_items = 0;
+  thrust::device_vector<int> out_num_runs(1, 42);
+
+  // Note intentionally no discard_iterator as we want to ensure nothing is written to the output arrays
+  run_length_encode(static_cast<int*>(nullptr),
+                    static_cast<int*>(nullptr),
+                    static_cast<int*>(nullptr),
+                    out_num_runs.begin(),
```

Comment:
suggestion:
```suggestion
                    thrust::raw_pointer_cast(out_num_runs.data()),
```

Reply:
What is the goal of that change?

Reply:
This relates to https://github.com/NVIDIA/cccl/pull/411#discussion_r1319732598. The idea is to reveal the issue with writing to device memory from the host. After this change, the test will fail.

Currently we are passing a fancy iterator for `d_num_runs_out`. That fancy iterator would detect that we are trying to write to device memory from the host and take precautions that we can successfully do so.